Migrate OwnerRRef value store to generic torch Future

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32608 Migrate OwnerRRef value store to generic torch Future**

It's a followup of https://github.com/pytorch/pytorch/pull/32556, where an error handling boilerplate code path was added to the FutureMessage callback.

However, I noticed that the FutureMessage could never be set with an error, because the FutureMessage is a member in OwnerRRef, 

- OwnerRRef does not have a setError method yet.
- The FutureMessage is only used for signaling
- The value of the RRef is contained in the `value_` field.

With the Future being generalized, it could contain more value types, not limited to Message.

This PR migrates the OwnerRRef value from the `value_` field to the generic Future.

In a later PR, it will be super easy to add a `setError` method for OwnerRRef, which calls `future_.setError(..)`. (I decide to do it later. I think it's better to migrate the call sites together with adding the new `setError` method.)

Differential Revision: [D5707692](https://our.internmc.facebook.com/intern/diff/D5707692/)

xush6528 committed:
Migrate OwnerRRef value store to generic torch Future

It's a followup of https://github.com/pytorch/pytorch/pull/32556.

Where an error handling boilerplate code path was added to the future callback

However, I noticed that the Future could never be set with an error, because the future is a member in OwnerRRef, while OwnnerRRef does not have a setError method yet.

Also, with the Future being generalized, it could .

This PR is a first step. It migrates the value store to the future owned by OwnerRRef, instead of having OwnerRRef to own a "value_" field.

In a later PR, it will be super easy to add a `setError` method for OwnerRRef, which calls `future.setError`. (I decide to do it later. I want to migrate the call sites tother with adding the new method.)

Differential Revision: [D5707692](https://our.internmc.facebook.com/intern/diff/D5707692/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2b1d9620:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2b1d962095c797a412c0ed779e4932eb57d0df18).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4339325) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90634433) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90634433))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
do so (now or later) by using -b with the checkout command again. Example: 
 
  git checkout -b <new-branch-name> 
 
HEAD is now at 2b1d962095 Migrate OwnerRRef value store to generic torch Future 
+ git reset --hard 2b1d962095c797a412c0ed779e4932eb57d0df18 
HEAD is now at 2b1d962095 Migrate OwnerRRef value store to generic torch Future 
+ git merge --allow-unrelated-histories --no-edit --no-ff e0ffe72649cef3a1dfd321545194d1b3574975c7 
Auto-merging torch/csrc/distributed/rpc/request_callback_impl.cpp 
CONFLICT (content): Merge conflict in torch/csrc/distributed/rpc/request_callback_impl.cpp 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "Migrate OwnerRRef value store to generic torch Future"


It's a followup of https://github.com/pytorch/pytorch/pull/32556.

Where an error handling boilerplate code path was added to the future callback

However, I noticed that the Future could never be set with an error, because the future is a member in OwnerRRef, while OwnnerRRef does not have a setError method yet.

Also, with the Future being generalized, it could .

This PR is a first step. It migrates the value store to the future owned by OwnerRRef, instead of having OwnerRRef to own a "value_" field.

In a later PR, it will be super easy to add a `setError` method for OwnerRRef, which calls `future.setError`. (I decide to do it later. I want to migrate the call sites tother with adding the new method.)

Differential Revision: [D5707692](https://our.internmc.facebook.com/intern/diff/D5707692/)

[ghstack-poisoned]

xush6528 committed:
Update on "Migrate OwnerRRef value store to generic torch Future"


It's a followup of https://github.com/pytorch/pytorch/pull/32556.

Where an error handling boilerplate code path was added to the future callback

However, I noticed that the Future could never be set with an error, because the future is a member in OwnerRRef, while OwnnerRRef does not have a setError method yet.

Also, with the Future being generalized, it could .

This PR is a first step. It migrates the value store to the future owned by OwnerRRef, instead of having OwnerRRef to own a "value_" field.

In a later PR, it will be super easy to add a `setError` method for OwnerRRef, which calls `future.setError`. (I decide to do it later. I want to migrate the call sites tother with adding the new method.)

Differential Revision: [D5707692](https://our.internmc.facebook.com/intern/diff/D5707692/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -167,40 +167,37 @@ IValue UserRRef::toHere() {
-  if (future.get() && !future->completed()) {
-    future->markCompleted(Message());
+  std::unique_lock<std::mutex> lock(futureValuePtrMutex_);
+  if (futureValuePtr_->completed()) {
```
The rest of this change seems fairly reasonable, but  this double-setValue() logic here seems to be potentially error-prone.

Do we need the ability to setValue() twice on an ownerRRef?

Creating a second future here seems to negate the value of having a Future<IValue> altogether. We can end up with problematic uses like the caller calling getFuture()->addCallback() and, if setValue is called twice, never being notified with the real value in some cases.

mrshenli replied:
> Do we need the ability to setValue() twice on an ownerRRef?

Based on the current design and API contract, yes, as RRef points to mutable data objects. 

I agree with @jjlilley's assessment here. When the future is already completed, `getFuture()` and the next `setValue()` could hit a race.

mrshenli replied:
It seems the only place that we call `RRef->getFuture` is in `request_callback_impl.cpp`, and purpose is to install a `callback` to the future. Will it be sufficient if we expose that `allCallback` API directly to RRef, make it atomic, and kill the `getFuture` API? Or is there any other potential use cases for `getFuture`?

xush6528 replied:
@jjlilley 

Yeah, it's like a dangling future, to which users can still add callbacks.

I think @mrshenli's solution is correct. This way, we hide future, and use OwnerRRef similarly to a future. Updated with this solution.

jjlilley replied:
It's usually a source of deadlocks to call something like markCompleted(), which may call callbacks, with the lock held.

The thing is, here, it's tricky to be correct here without the lock held.

Would we be better off instead just adding an explicit error+accessors to this class, rather than using Future<> in this regard?
It feels like we're trying to shoe-horn in the Future<IValue> here, with a pattern that doesn't quite fit.

xush6528 committed:
Update on "Migrate OwnerRRef value store to generic torch Future"


It's a followup of https://github.com/pytorch/pytorch/pull/32556, where an error handling boilerplate code path was added to the FutureMessage callback.

However, I noticed that the FutureMessage could never be set with an error, because the FutureMessage is a member in OwnerRRef, 

- OwnerRRef does not have a setError method yet.
- The FutureMessage is only used for signaling
- The value of the RRef is contained in the `value_` field.

With the Future being generalized, it could contain more value types, not limited to Message.

This PR migrates the OwnerRRef value from the `value_` field to the generic Future.

In a later PR, it will be super easy to add a `setError` method for OwnerRRef, which calls `future_.setError(..)`. (I decide to do it later. I think it's better to migrate the call sites together with adding the new `setError` method.)

Differential Revision: [D5707692](https://our.internmc.facebook.com/intern/diff/D5707692/)

[ghstack-poisoned]
